cpp/src/arrow/filesystem/azurefs.cc

```diff
@@ -1662,20 +1705,28 @@ class AzureFileSystem::Impl {
       AzureFileSystem* fs) {
     RETURN_NOT_OK(ValidateFileLocation(location));
 
+    const auto blob_container_client = GetBlobContainerClient(location.container);
     auto block_blob_client = std::make_shared<Blobs::BlockBlobClient>(
-        blob_service_client_->GetBlobContainerClient(location.container)
-            .GetBlockBlobClient(location.path));
+        blob_container_client.GetBlockBlobClient(location.path));
+
+    auto ensure_not_flat_namespace_directory = [this, location,
+                                                blob_container_client]() -> Status {
+      bool hierarchical_namespace_enabled =
+          HierarchicalNamespaceSupport(GetFileSystemClient(location.container)) ==
+          HNSSupport::kEnabled;
```

Comment:
It could be `HNSSupport::kContainerNotFound`.

You can be explicit

```cpp
    ARROW_ASSIGN_OR_RAISE(auto hns_support, HierarchicalNamespaceSupport(adlfs_client));
    if (hns_support == HNSSupport::kContainerNotFound) {
      return Status::OK();
    }
    if (hns_support == HNSSupport::kDisabled) {
      // handle HNS disabled case
    }
    return Status::OK();
```

or compare with `HNSSupport::kDisabled` (`!(hns_suppport == kEnabled)` is not equivalent to `hns_support == kDisabled`).

```cpp
    ARROW_ASSIGN_OR_RAISE(auto hns_support, HierarchicalNamespaceSupport(adlfs_client));
    if (hns_support == HNSSupport::kDisabled) {
       // check
    }
    return Status::OK();
```

Reply:
:+1:  we can save ourselves an extra API call if its `HNSSupport::kContainerNotFound`